pkg/operator/ceph/nvmeof/controller.go

```diff
@@ -0,0 +1,624 @@
+	APIVersion: fmt.Sprintf("%s/%s", cephv1.CustomResourceGroup, cephv1.Version),
+}
+
+var currentAndDesiredCephVersion = opcontroller.CurrentAndDesiredCephVersion
+
+type ReconcileCephNVMeOFGateway struct {
+	client                client.Client
+	scheme                *runtime.Scheme
+	context               *clusterd.Context
+	cephClusterSpec       *cephv1.ClusterSpec
+	clusterInfo           *cephclient.ClusterInfo
+	opManagerContext      context.Context
+	opConfig              opcontroller.OperatorConfig
+	recorder              record.EventRecorder
+	shouldRotateCephxKeys bool
+}
+
+// Add creates a new CephNVMeOFGateway Controller and adds it to the Manager.
+func Add(mgr manager.Manager, context *clusterd.Context, opManagerContext context.Context, opConfig opcontroller.OperatorConfig) error {
+	return add(mgr, newReconciler(mgr, context, opManagerContext, opConfig))
+}
+
+func newReconciler(mgr manager.Manager, context *clusterd.Context, opManagerContext context.Context, opConfig opcontroller.OperatorConfig) reconcile.Reconciler {
+	return &ReconcileCephNVMeOFGateway{
+		client:           mgr.GetClient(),
+		scheme:           mgr.GetScheme(),
+		context:          context,
+		opManagerContext: opManagerContext,
+		opConfig:         opConfig,
+		recorder:         mgr.GetEventRecorderFor("rook-" + controllerName),
+	}
+}
+
+func watchOwnedCoreObject[T client.Object](c controller.Controller, mgr manager.Manager, obj T) error {
+	return c.Watch(
+		source.Kind(
+			mgr.GetCache(),
+			obj,
+			handler.TypedEnqueueRequestForOwner[T](
+				mgr.GetScheme(),
+				mgr.GetRESTMapper(),
+				&cephv1.CephNVMeOFGateway{},
+			),
+			opcontroller.WatchPredicateForNonCRDObject[T](&cephv1.CephNVMeOFGateway{TypeMeta: controllerTypeMeta}, mgr.GetScheme()),
+		),
+	)
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		logger.Errorf("failed to create controller: %v", err)
+		return err
+	}
+	logger.Info("successfully started")
+
+	err = addonsv1alpha1.AddToScheme(mgr.GetScheme())
+	if err != nil {
+		logger.Errorf("failed to add addonsv1alpha1 to scheme: %v", err)
+		return err
+	}
+
+	err = csiopv1.AddToScheme(mgr.GetScheme())
+	if err != nil {
+		logger.Errorf("failed to add csiopv1 to scheme: %v", err)
+		return err
+	}
+
+	err = c.Watch(
+		source.Kind(
+			mgr.GetCache(),
+			&cephv1.CephNVMeOFGateway{TypeMeta: controllerTypeMeta},
+			&handler.TypedEnqueueRequestForObject[*cephv1.CephNVMeOFGateway]{},
+			opcontroller.WatchControllerPredicate[*cephv1.CephNVMeOFGateway](mgr.GetScheme()),
+		),
+	)
+	if err != nil {
+		logger.Errorf("failed to watch CephNVMeOFGateway CRD: %v", err)
+		return err
+	}
+
+	for _, t := range objectsToWatch {
+		err = watchOwnedCoreObject(c, mgr, t)
+		if err != nil {
+			logger.Errorf("failed to watch resource type %s: %v", t.GetObjectKind().GroupVersionKind().String(), err)
```

Comment:
No need to log an error when returning the error. Return a wrapped error as needed.